File: src/utils/calculator/getAttrData.ts
import officalLightconeId from "../../../map/lightcone_offical_id_map";
import { CharacterName } from "../../types/character";
import { LightconeName } from "../../types/lightcone";
import { getCharFullData, getLcFullData } from "../data/getDataFromMap";

const levelMin = [1, 20, 30, 40, 50, 60, 70];
const levelMax = [20, 30, 40, 50, 60, 70, 80];

//* 根據等級取得角色屬性數值
export function getCharAttrData(id: CharacterName, level: number = 1) {
  const charFullData = getCharFullData(id);
  const charLevelData = charFullData?.levelData;

  const attributes = {
    atk: 0,
    hp: 0,
    def: 0,
    speed: 0,
    aggro: 0,
    energy: charFullData?.spRequirement,
  };

  // 找到對應等級的數據
  const dataForLevel = charLevelData?.find((data) =>
    level === 80 ? level <= data?.maxLevel : level < data?.maxLevel
  );

  if (dataForLevel) {
    // 計算屬性值
    attributes.atk =
      dataForLevel.attackBase + dataForLevel.attackAdd * (level - 1);
    attributes.hp = dataForLevel.hpBase + dataForLevel.hpAdd * (level - 1);
    attributes.def =
      dataForLevel.defenseBase + dataForLevel.defenseAdd * (level - 1);
    attributes.speed =
      dataForLevel.speedBase + dataForLevel.speedAdd * (level - 1);
    attributes.aggro = dataForLevel.aggro;
  }

  return attributes;
}

//* 根據等級取得角色屬性數值
export function getLcAttrData(id: LightconeName, level: number = 1) {
  const lcFullData = getLcFullData(id);
  const lcLevelData = lcFullData.levelData;

  const attributes = {
    atk: 0,
    hp: 0,
    def: 0,
  };

  // 找到對應等級的數據
  const dataForLevel = lcLevelData.find((data) => level === 80 ? level <= data.maxLevel : level < data.maxLevel);

  if (dataForLevel) {
    // 計算屬性值
    attributes.atk =
      dataForLevel.attackBase + dataForLevel.attackAdd * (level - 1);
    attributes.hp = dataForLevel.hpBase + dataForLevel.hpAdd * (level - 1);
    attributes.def =
      dataForLevel.defenseBase + dataForLevel.defenseAdd * (level - 1);
  }

  return attributes;
}

//* 根據等級取得角色屬性數值
export function getLcAttrDataJSON(id: number, level: number = 1) {
  const lcFullData = getLcFullData(officalLightconeId[id]);
  const lcLevelData = lcFullData.levelData;

  const attributes = [
    {
      field: "hp",
      value: 0,
      percent : false,
      display : "0.0",
    },
    {
      field: "hp",
      value: 0,
      percent : false,
      display : "0.0",
    },
    {
      field: "def",
      value: 0,
      percent : false,
      display : "0.0",
    },
  ];

  // 找到對應等級的數據
  const dataForLevel = lcLevelData.find((data) => level === 80 ? level <= data.maxLevel : level < data.maxLevel) ;

  if (dataForLevel) {
    // 計算屬性值
    attributes[0].value =
      dataForLevel.attackBase + dataForLevel.attackAdd * (level - 1);
    attributes[1].value = dataForLevel.hpBase + dataForLevel.hpAdd * (level - 1);
    attributes[2].value =
      dataForLevel.defenseBase + dataForLevel.defenseAdd * (level - 1);

      attributes.map((attribute : any) => {
        attribute.display = (attribute.value * (attribute.percent ? 100 : 1)).toFixed(1).toString()+(attribute.percent ? "%" : "")
      })
  }

  return attributes;
}

File: composeApp/src/commonMain/kotlin/utils/calculator/GetAttrData.kt
package utils.calculator

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.float
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * This file is ref from https://github.com/Coding-band/Honkai-Stargazer/blob/rn_branch/src/utils/calculator/getAttrData.ts
 * (寫算法係我自己吖嘛...拎嚟用返冇問題掛...)
 */
@Serializable
data class AttrData(
    var atk: Float,
    var hp: Float,
    var def: Float,
    var spd : Float,
    var aggro: Int,
    var energy: Int,
)
// 根據等級取得角色屬性數值
@OptIn(ExperimentalCoroutinesApi::class)
fun getCharAttrData(jsonElement: JsonElement, level: Int = 1) : AttrData{
    return runBlocking {
        val job = CoroutineScope(Dispatchers.Default).async {
            val charLevelData = jsonElement.jsonObject["levelData"]!!
            var tmpAttrData : AttrData = AttrData(0f,0f,0f,0f,0,0)

            // 找到對應等級的數據
            val dataFromLevel = charLevelData.jsonArray.find { data ->
                if(level == 80) {
                    level <= data.jsonObject["maxLevel"]!!.jsonPrimitive.int
                } else {
                    level < data.jsonObject["maxLevel"]!!.jsonPrimitive.int
                }
            }

            if(dataFromLevel !== null){
                tmpAttrData.atk = dataFromLevel.jsonObject["attackBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["attackAdd"]!!.jsonPrimitive.float) * (level - 1)
                tmpAttrData.def = dataFromLevel.jsonObject["defenseBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["defenseAdd"]!!.jsonPrimitive.float) * (level - 1)
                tmpAttrData.hp = dataFromLevel.jsonObject["hpBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["hpAdd"]!!.jsonPrimitive.float) * (level - 1)
                tmpAttrData.spd = dataFromLevel.jsonObject["speedBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["speedAdd"]!!.jsonPrimitive.float) * (level - 1)
                tmpAttrData.aggro = dataFromLevel.jsonObject["aggro"]!!.jsonPrimitive.int
                tmpAttrData.energy = jsonElement.jsonObject["spRequirement"]!!.jsonPrimitive.int
            }

            return@async tmpAttrData
        }
        job.await()
        job.getCompleted()
    }
}



// 根據等級取得光錐屬性數值
@OptIn(ExperimentalCoroutinesApi::class)
fun getLcAttrData(jsonElement: JsonElement, level: Int = 1) : AttrData{
    return runBlocking {
        val job = async (Dispatchers.Default){
            val charLevelData = jsonElement.jsonObject["levelData"]!!
            var tmpAttrData : AttrData = AttrData(0f,0f,0f,0f,0,0)

            // 找到對應等級的數據
            val dataFromLevel = charLevelData.jsonArray.find { data ->
                if(level == 80) {
                    level <= data.jsonObject["maxLevel"]!!.jsonPrimitive.int
                } else {
                    level < data.jsonObject["maxLevel"]!!.jsonPrimitive.int
                }
            }

            if(dataFromLevel !== null){
                tmpAttrData.atk = dataFromLevel.jsonObject["attackBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["attackAdd"]!!.jsonPrimitive.float) * (level - 1)
                tmpAttrData.def = dataFromLevel.jsonObject["defenseBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["defenseAdd"]!!.jsonPrimitive.float) * (level - 1)
                tmpAttrData.hp = dataFromLevel.jsonObject["hpBase"]!!.jsonPrimitive.float + (dataFromLevel.jsonObject["hpAdd"]!!.jsonPrimitive.float) * (level - 1)
            }

            return@async  tmpAttrData
        }
        job.await()
        job.getCompleted()
    }
}

